bin/propolis-server/src/lib/stats/mod.rs

```diff
@@ -217,3 +239,149 @@ pub(crate) async fn track_vcpu_kstats(
+        NETWORK_INTERFACE_SAMPLE_INTERVAL,
+    );
+    let interface_id = nics.target.interface_id;
+    if let Err(e) = sampler.add_target(nics, details).await {
+        slog::error!(
+            log,
+            "failed to add network interface targets, \
+            network interface stats will be unavailable";
+            "network_interface_id" => %interface_id,
+            "error" => ?e,
+        );
+    }
+}
+
+#[cfg(all(not(test), target_os = "illumos"))]
+mod kstat_types {
+    pub(crate) use kstat_rs::{Data, Kstat, Named, NamedData};
+    pub(crate) use oximeter_instruments::kstat::{
+        hrtime_to_utc, ConvertNamedData, Error, KstatList, KstatTarget,
+    };
+}
+
+/// Mock the relevant subset of `kstat-rs` types needed for tests.
+#[cfg(not(all(not(test), target_os = "illumos")))]
+#[allow(dead_code, unused)]
+mod kstat_types {
+    use chrono::DateTime;
+    use chrono::Utc;
+    use oximeter::{Sample, Target};
+
+    pub(crate) type KstatList<'a, 'k> =
+        &'a [(DateTime<Utc>, Kstat<'k>, Data<'k>)];
+
+    pub(crate) trait KstatTarget:
+        Target + Send + Sync + 'static + std::fmt::Debug
+    {
+        /// Return true for any kstat you're interested in.
+        fn interested(&self, kstat: &Kstat<'_>) -> bool;
+
+        /// Convert from a kstat and its data to a list of samples.
+        fn to_samples(
+            &self,
+            kstats: KstatList<'_, '_>,
+        ) -> Result<Vec<Sample>, Error>;
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) enum Data<'a> {
+        Named(Vec<Named<'a>>),
+        Null,
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) enum NamedData<'a> {
+        UInt32(u32),
+        UInt64(u64),
+        String(&'a str),
+    }
+
+    #[derive(Debug)]
+    pub(crate) struct Kstat<'a> {
+        pub ks_module: &'a str,
+        pub ks_instance: i32,
+        pub ks_name: &'a str,
+        pub ks_snaptime: i64,
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) struct Named<'a> {
+        pub name: &'a str,
+        pub value: NamedData<'a>,
+    }
+
+    #[allow(unused)]
+    pub(crate) trait ConvertNamedData {
+        fn as_i32(&self) -> Result<i32, Error>;
+        fn as_u32(&self) -> Result<u32, Error>;
+        fn as_i64(&self) -> Result<i64, Error>;
+        fn as_u64(&self) -> Result<u64, Error>;
+    }
+
+    impl<'a> ConvertNamedData for NamedData<'a> {
+        fn as_i32(&self) -> Result<i32, Error> {
+            unimplemented!()
+        }
+
+        fn as_u32(&self) -> Result<u32, Error> {
+            if let NamedData::UInt32(x) = self {
+                Ok(*x)
+            } else {
+                panic!()
+            }
+        }
+
+        fn as_i64(&self) -> Result<i64, Error> {
+            unimplemented!()
+        }
+
+        fn as_u64(&self) -> Result<u64, Error> {
+            if let NamedData::UInt64(x) = self {
+                Ok(*x)
+            } else {
+                panic!()
```

Comment:
Why `panic` instead of emitting an error?

This is currently causing `stats::network_interface::test::test_kstat_to_samples` to fail on non-illumos

Reply:
That was leftover from the previous code, but will update.

Reply:
And, also update the test :).

Reply:
The test may pass again if some of this shim code is updated to throw errors rather than panic.

Reply:
The `panic!()` was intentional. It looks like the test you're referring to is failing because the code converting from the kstat to the oximeter sample is trying to convert a `NamedData::U32(_)` to a `u64`. We can return an error, but the code is wrong in either case. I wanted to loudly, quickly, and irrecoverably when testing conversion routines that expect some relationship between the types of the kstat and the oximeter sample it generates, in this case, that they're both `u64`s under the hood.

Reply:
Yep. I quickly fixed up the code after changing NamedData type and moved to the error anyhow.